### ISSUE TITLE
`kernel_size` can be zero, giving nonsense results, when enough distances are zero

### ISSUE THREAD

**Issue by Linux-cpp-lisp** [cwehmeyer/pydpc#2]
Hi,

First off, thanks for making this optimized DPC implementation available -- it's a big help.

When a sufficient (`fraction * n`) quantity of off-diagonal zeros are present in the pairwise distance matrix, the kernel size can be set to zero, leading to division by zero and nonsense zeros and NaNs in the density. 

This could be solved relatively simply by adding a fallback case to [`_get_kernel_size()`](https://github.com/cwehmeyer/pydpc/blob/master/ext/_core.c#L114):
```c
// ...
if(kernel_size == 0.0) {
    for(int i = 0; i < n; i++) {
        if(scratch[i] != 0.0) {
            kernel_size = scratch[i];
            break;
        }
    }
}
```

Taking the first non-zero distance as the width of the Gaussian seems like a reasonable fallback assumption. Under any circumstances, if `pydpc` doesn't want to presume that this is a reasonable fallback, either the user should be given the ability to specify `kernel_width` or an error should be thrown.

(I don't think this scenario is too pathological, either: I came across it because I had duplicate points under the Euclidean metric, which may or may not make sense depending on the problem, but a user could also pass in a distance matrix generated by some function that, for example, marks as identical inputs within some cutoff, etc., etc.)

Thanks,
A.M.

**Comment by Linux-cpp-lisp** [comment]
(I'm more than happy to implement a fix and file a PR; I just felt that it was worth discussing which solution would be most appropriate first.)

**Comment by quantkeyvis** [comment]
I haven't gotten any updates from the devs but if they are adding fixes I'd
add one to choose the distance metric. I can easily add it to the repo
myself of they give me permission to

**Comment by Linux-cpp-lisp** [comment]
@quantkeyvis (Not really relevant to this issue, but either way.) In the sense of implementing other norms in this package, or in the sense of allowing a custom pairwise distance matrix to be passed? I'd lean to the second to allow, for example, any of `sklearn`'s efficient metrics to be used.

Have you filed a PR? The last commit on this repo by the owner was in April of this year; it seems to still be maintained.

**Comment by cwehmeyer** [comment]
Hi @Linux-cpp-lisp,

thank you for raising this issue!

Your proposed fallback is a reasonable solution for this problem. I further agree that a user should be able to specify the kernel size manually.

If you already have the fix in your fork, I would appreciate a pull request.

**Comment by cwehmeyer** [comment]
Hi @quantkeyvis,

having the choice of distance metric is a nice idea, thanks! Feel free to file a pull request if you have the code available.

**Comment by Linux-cpp-lisp** [comment]
@cwehmeyer  wanted to wait and get your thoughts, but I can code that up quickly and file a PR :+1:

**Comment by Linux-cpp-lisp** [comment]
@cwehmeyer  PR filed (#3), let me know what you think.

@quantkeyvis Turns out, as I was going back into my fork to fix this, I already had code to allow a user-provided distance matrix, so I've filled a second PR (#4) with that code. Does that look like what you were thinking of?

**Comment by quantkeyvis** [comment]
I'll have to look through my code to see. I made a few other changes, one
of which is allowing the program to reduce the number of parameters by
multiplying the density by the nearest distance as they did in the original
paper.

**Comment by Linux-cpp-lisp** [comment]
@quantkeyvis to autoselect the thresholds? (Gamma, shown in Fig 4B in the paper.)

I also have some code (likely different from yours) for finding the elbow in that plot and automatically choosing the clustering thresholds from that. Maybe we ought to open a separate issue and discuss this?

**Comment by quantkeyvis** [comment]
My only suggestion it to multiply the two metrics as they did in the paper.
The exact cutoff point can change depending on the data so I'd leave that
up to the user. I do expect users of analytics to have some understanding
of the underlying maths.

**Comment by Linux-cpp-lisp** [comment]
I don't see why the library should provide gamma (the multiplied metric) if it isn't going to do anything with it... at that point, I think the user is more than capable of doing the computation themselves if they need it.

**Comment by quantkeyvis** [comment]
1. Its cumbersome to do on your own

2. The cutoff metric doesn't  necessarily convert to the two metrics in a
way that selects only the proper points

3. The package claims to reproduce the paper, since this is a crucial
metric at least it should be stated that the package has this distance.

**Comment by quantkeyvis** [comment]
4. It would reduce the number of parameters

**Comment by Linux-cpp-lisp** [comment]
Resolved by #3
